File: DevEnvironment/Command/src/main/java/command/core/Parameter.java
package command.core;

import base.core.Conditions;
import base.core.Conversion;
import base.core.IFromString;
import command.interfaces.IParameter;
import command.interfaces.IParameterMetadata;
import java.math.BigInteger;
import java.time.Duration;
import java.util.Date;

/**
 * The Parameter class implements a parameter.
 */
public final class Parameter implements IParameter {
    private final IParameterMetadata metaData;
    private final String value;

    /**
     * Creates a help parameter.
     */
    public static IParameter createHelpParameter() {
        return new Parameter(ParameterMetadata.createHelpParameter(), null);
    }

    /**
     * The Parameter constructor.
     */
    public Parameter(
        IParameterMetadata metaData,
        String value) {

        Conditions.validateNotNull(
            metaData,
            "The metaData of a parameter.");

        this.metaData = metaData;
        this.value = value;
    }

    /**
     * Gets a metadata of a parameter.
     */
    @Override
    public IParameterMetadata getMetadata() {
        return this.metaData;
    }

    /**
     * Checks whether the value of the parameter is set.
     */
    @Override
    public boolean isSet() {
        return this.value != null;
    }

    /**
     * Gets a boolean value of a parameter.
     */
    @Override
    public boolean getBooleanValue() {
        return Conversion.stringConversion().toBoolean(this.value);
    }

    /**
     * Gets a byte value of a parameter.
     */
    @Override
    public byte getByteValue() {
        return Conversion.stringConversion().toByte(this.value);
    }

    /**
     * Gets a short value of a parameter.
     */
    @Override
    public short getShortValue() {
        return Conversion.stringConversion().toShort(this.value);
    }

    /**
     * Gets an integer value of a parameter.
     */
    @Override
    public int getIntegerValue() {
        return Conversion.stringConversion().toInteger(this.value);
    }

    /**
     * Gets a long value of a parameter.
     */
    @Override
    public long getLongValue() {
        return Conversion.stringConversion().toLong(this.value);
    }

    /**
     * Gets a float value of a parameter.
     */
    @Override
    public float getFloatValue() {
        return Conversion.stringConversion().toFloat(this.value);
    }

    /**
     * Gets a double value of a parameter.
     */
    @Override
    public double getDoubleValue() {
        return Conversion.stringConversion().toDouble(this.value);
    }

    /**
     * Gets a character value of a parameter.
     */
    @Override
    public char getCharacterValue() {
        return Conversion.stringConversion().toCharacter(this.value);
    }

    /**
     * Gets a string value of a parameter.
     */
    @Override
    public String getStringValue() {
        return this.value;
    }

    /**
     * Gets an enum value of a parameter.
     */
    @Override
    public <T extends Enum<T>> T getEnumValue(IFromString<T> transformer) {
        return transformer.fromString(this.value);
    }

    /**
     * Gets a big integer value of a parameter.
     */
    @Override
    public BigInteger getBigIntegerValue(int radix) {
        return Conversion.stringConversion().toBigInteger(this.value, radix);
    }

    /**
     * Gets a date value of a parameter.
     */
    @Override
    public Date getDateValue() {
        return Conversion.stringConversion().toDate(this.value);
    }

    /**
     * Gets a duration value of a parameter.
     */
    @Override
    public Duration getDurationValue() {
        return Conversion.stringConversion().toDuration(this.value);
    }

    /**
     * Gets a string array value of a parameter.
     */
    @Override
    public String[] getStringArrayValue() {
        String value = this.value;

        if (value == null) {
            return null;
        }

        return value.split(",");
    }
}

File: DevEnvironmentCLion/Command/core/Parameter.cpp
#include "PreCompiled.h"
#include "Parameter.h"
#include "CommandConstants.h"
#include "Conversion.h"
#include "Strings.h"

using namespace command;

/**
 * Creates a new parameter.
 */
IParameterSharedPtr Parameter::make(
    const std::string& name,
    const std::string& value,
    bool defined)
{
    return std::make_shared<Parameter>(name, value, defined);
}

/**
 * The Parameter constructor.
 */
Parameter::Parameter(
    const std::string& name,
    const std::string& value,
    bool defined) :
    m_name(name),
    m_value(value),
    m_defined(defined)
{
}

/**
 * The Parameter destructor.
 */
Parameter::~Parameter()
{
}

/**
 * Gets the name of the parameter.
 */
const std::string& Parameter::getName() const
{
    return m_name;
}

/**
 * Checks whether the value of the parameter is defined.
 */
bool Parameter::isDefined() const
{
    return m_defined;
}

/**
 * Gets a boolean value of a parameter.
 */
bool Parameter::getBooleanValue() const
{
    return base::Conversion::stringConversion().toBoolean(m_value);
}

/**
 * Gets a short value of a parameter.
 */
short Parameter::getShortValue() const
{
    return base::Conversion::stringConversion().toShort(m_value);
}

/**
 * Gets an integer value of a parameter.
 */
int Parameter::getIntegerValue() const
{
    return base::Conversion::stringConversion().toInteger(m_value);
}

/**
 * Gets a long value of a parameter.
 */
long Parameter::getLongValue() const
{
    return base::Conversion::stringConversion().toLong(m_value);
}

/**
 * Gets a size_t value of a parameter.
 */
size_t Parameter::getSizeTValue() const
{
    return base::Conversion::stringConversion().toSizeT(m_value);
}

/**
 * Gets a float value of a parameter.
 */
float Parameter::getFloatValue() const
{
    return base::Conversion::stringConversion().toFloat(m_value);
}

/**
 * Gets a double value of a parameter.
 */
double Parameter::getDoubleValue() const
{
    return base::Conversion::stringConversion().toDouble(m_value);
}

/**
 * Gets a character value of a parameter.
 */
char Parameter::getCharacterValue() const
{
    return base::Conversion::stringConversion().toCharacter(m_value);
}

/**
 * Gets a string value of a parameter.
 */
const std::string& Parameter::getStringValue() const
{
    return m_value;
}

/**
 * Gets a duration value of a parameter.
 */
base::DurationSharedPtr Parameter::getDurationValue() const
{
    return base::Conversion::stringConversion().toDuration(m_value);
}

/**
 * Gets a string array value of a parameter.
 */
void Parameter::getStringArrayValue(std::vector<std::string>& result) const
{
    base::Strings::splitString(m_value, ",", result);
}
